core/src/main/scala/cats/CommutativeApplicative.scala

```diff
@@ -0,0 +1,36 @@
+package cats
+
+import simulacrum.typeclass
+
+/**
+  * Commutative Applicative.
+  *
+  * Further than an Applicative, which just allows composition of independent effectful functions,
+  * in a Commutative Applicative those functions can be composed in any order, which guarantees
+  * that their effects do not interfere.
+  *
+  * Must obey the laws defined in cats.laws.CommutativeApplicativeLaws.
+  */
+@typeclass trait CommutativeApplicative[F[_]] extends Applicative[F] with CommutativeApply[F] {
```

Comment:
I wonder if we really want an `UnorderedFoldable` and `UnorderedTraverse` that instead of `toList` has `toSet` and has a `foldMap` that takes a `CommutativeMonoid` and for `sequence/traverse` uses `CommutativeApplicative`.

That seems cleaner to me than putting these methods here now that I think about it.

I think we can construct sane laws for `UnorderedFoldable` that would make sense. If we added a method like `insert[T](f: F[T], item: T): F[T]` then I think you could make some law.

Reply:
I agree that an extra type class would feel a lot cleaner than baking these 4 operations into the `CommutativeApplicative` type class. Related thoughts [here](https://github.com/typelevel/cats/issues/1831#issuecomment-337221498).
I've played around with some laws, and we wouldn't have to get rid of too many laws too make it fit. I also like the `insert`, although I'd have to think about it more. We could let `Foldable` extend `UnorderedFoldable`, if that makes any sense. I'll play around with it and update the PR accordingly :)

Reply:
This opens up a can of worms though. We could probably also define something like `unorderedFoldM` with `CommutativeMonad`, a `UnorderedReducible`, `NonEmptyUnordered` and all kinds of others. I came up with something crazy a while ago on gitter https://gitter.im/typelevel/cats-dev?at=59cf81a5210ac269208e0bcf but seems I've only scratched the surface :smile: 

I'll try to define this as minimal as possible and go from there. Might make a good blog post though.

Reply:
I had to remove `insert` for now, because it's not possible to implement it for `Map` :/